### PR TITLE
New data source `azurerm_mssql_server` and new block `restorable_dropped_databases` in `azuerrm_mssql_server`

### PR DESCRIPTION
**Pull request by yupwei68**
Fix https://github.com/terraform-providers/terraform-provider-azurerm/issues/6502
https://github.com/terraform-providers/terraform-provider-azurerm/issues/7594

=== RUN   TestAccAzureRMMsSqlDatabase_createRestoreMode
=== PAUSE TestAccAzureRMMsSqlDatabase_createRestoreMode
=== CONT  TestAccAzureRMMsSqlDatabase_createRestoreMode
--- PASS: TestAccAzureRMMsSqlDatabase_createRestoreMode (2016.02s)

Sql server listing recoverableDatabases Api has problems, thus it's excluded in the PR: https://github.com/Azure/azure-rest-api-specs/issues/10162

### CODE REVIEW

File: website/docs/d/mssql_server.html.markdown [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+* `administrator_login` - The administrator login name for the new server.
+
+* `fully_qualified_domain_name` - The fully qualified domain name of the Azure SQL Server.
+
+* `identity` - A `identity` block as defined below.
+
+* `location` - The Azure Region where the Microsoft SQL Server exists.
+
+* `restorable_dropped_databases` - A list of restorable database IDs of the server.
```
Comment (katbyte) [inline website/docs/d/mssql_server.html.markdown:48]:
```suggestion
* `restorable_dropped_databases` - A list of dropped restorable database IDs on the server.
```

File: website/docs/d/mssql_server.html.markdown [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+* `tags` - A mapping of tags assigned to the Microsoft SQL Server.
```
Comment (katbyte) [inline website/docs/d/mssql_server.html.markdown:50]:
```suggestion
* `tags` - A mapping of tags assigned to this Microsoft SQL Server.
```

File: website/docs/d/mssql_server.html.markdown [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+* `version` - The version for the server.
```
Comment (katbyte) [inline website/docs/d/mssql_server.html.markdown:52]:
```suggestion
* `version` - This servers MS SQL version.
```

File: website/docs/r/mssql_server.html.markdown [hunk 4]
```diff
@@ -99,6 +99,8 @@ The following attributes are exported:
 
 * `fully_qualified_domain_name` - The fully qualified domain name of the Azure SQL Server (e.g. myServerName.database.windows.net)
 
+* `restorable_dropped_databases` - A list of restorable database IDs of the server.
```
Comment (katbyte) [inline website/docs/r/mssql_server.html.markdown:102]:
```suggestion
* `restorable_dropped_databases` - A list of dropped restorable database IDs on the server.
```

File: website/docs/d/mssql_server.html.markdown [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+* `resource_group_name` - (Required) The name of the Resource Group where the Microsoft SQL Server exists.
+
+## Attributes Reference
+
+In addition to the Arguments listed above - the following Attributes are exported: 
+
+* `id` - The ID of the Microsoft SQL Server.
+
+* `administrator_login` - The administrator login name for the new server.
```
Comment (katbyte) [inline website/docs/d/mssql_server.html.markdown:40]:
```suggestion
* `administrator_login` - The administrator's login name for the new server.
```

File: azurerm/internal/services/mssql/mssql_database_resource.go [hunk 6]
```diff
@@ -134,6 +134,16 @@ func resourceArmMsSqlDatabase() *schema.Resource {
 				ValidateFunc:     validation.IsRFC3339Time,
 			},
 
+			"recoverable_database_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"restorable_dropped_database_id": {
```
Comment (katbyte) [inline azurerm/internal/services/mssql/mssql_database_resource.go:142]: Would this make more sense as
```suggestion
			"recover_dropped_database_id": {
```

File: azurerm/internal/services/mssql/mssql_database_resource.go [hunk 7]
```diff
@@ -134,6 +134,16 @@ func resourceArmMsSqlDatabase() *schema.Resource {
 				ValidateFunc:     validation.IsRFC3339Time,
 			},
 
+			"recoverable_database_id": {
```
Comment (katbyte) [inline azurerm/internal/services/mssql/mssql_database_resource.go:137]: How does this work in relation to `restorable_dropped_database_id` ? they seem to be the same and i don't see a test for this?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the PR @yupwei68 - i've left some comments inline that need to be addressed before merge

File: azurerm/internal/services/mssql/mssql_database_resource.go (hunk 6 above)
Comment (yupwei68) [reply azurerm/internal/services/mssql/mssql_database_resource.go:142]: `recoverable` and `restorable` databases are two kinds of database, they are restored by `restore` and `recovery` two create mode. The difference, from my aspect, is that if you have deleted a db, you could immediate `restore` it. If it's too long time to be got from `restorable` dbs, you could get it from `recoverable` dbs and `recovery` it. I'd like to suggest not to change the name to this to cause confusions.

File: azurerm/internal/services/mssql/mssql_database_resource.go (hunk 7 above)
Comment (yupwei68) [reply azurerm/internal/services/mssql/mssql_database_resource.go:137]: the same above

**Comment by yupwei68** [comment]
Hi @katbyte Thanks for your comments. I've replied the unchanged comments, others have been pushed. Please continue reviewing.

File: azurerm/internal/services/mssql/mssql_database_resource.go (hunk 6 above)
Comment (katbyte) [reply azurerm/internal/services/mssql/mssql_database_resource.go:142]: ahh ok that makes sense, then this should be
```suggestion
			"restore_dropped_database_id": {
```

File: azurerm/internal/services/mssql/mssql_database_resource.go (hunk 7 above)
Comment (katbyte) [reply azurerm/internal/services/mssql/mssql_database_resource.go:137]: and then this 
```suggestion
			"recover_database_id": {
```
?

**Comment by yupwei68** [comment]
Hi @katbyte Thanks for your comments. Changes have been pushed. Please continue reviewing.

File: azurerm/internal/services/mssql/mssql_database_resource.go (hunk 6 above)
Comment (katbyte) [reply azurerm/internal/services/mssql/mssql_database_resource.go:142]: This still hasn't been changed to 
```suggestion
			"restore_dropped_database_id": {
```

**Review by katbyte — CHANGES_REQUESTED** [review]

**Comment by yupwei68** [comment]
@katbyte Sorry for the mistakes. Please continue reviewing.

File: azurerm/internal/services/mssql/mssql_server_resource.go [hunk 8]
```diff
@@ -150,6 +150,14 @@ func resourceArmMsSqlServer() *schema.Resource {
 				Computed: true,
 			},
 
+			"restore_dropped_database_id": {
```
Comment (jackofallops) [inline azurerm/internal/services/mssql/mssql_server_resource.go:153]: I realise there's already been some discussion on this, but I think the original `restorable_dropped_database_ids` is possibly more appropriate here as this computed field exposes those databases that have been dropped and are capable of being restored?

File: azurerm/internal/services/mssql/mssql_server_data_source.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+						},
+						"tenant_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+
+			"restore_dropped_database_id": {
```
Comment (jackofallops) [inline azurerm/internal/services/mssql/mssql_server_data_source.go:70]: As with the resource, I think `restorable_dropped_database_ids` is a more appropriate descriptive name as it is the possible source list of ID's for the `azurerm_mssql_database` resource to consume.

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @yupwei68 
A couple of additional comments below.  And apologies for re-opening the discussion on the property name, I'll try to catch up with @katbyte offline later today on that one.

Thanks

File: azurerm/internal/services/mssql/mssql_server_resource.go (hunk 8 above)
Comment (yupwei68) [reply azurerm/internal/services/mssql/mssql_server_resource.go:153]: I've changed the   `recover_database_id` to `recoverable_database_id`, too. As here, we could only apply one db, I've removed the plurality。

File: azurerm/internal/services/mssql/mssql_database_resource.go [hunk 10]
```diff
@@ -134,6 +134,18 @@ func resourceArmMsSqlDatabase() *schema.Resource {
...
```
Comment (katbyte) [inline azurerm/internal/services/mssql/mssql_database_resource.go:137]: Again, could we make this
```suggestion
			"recover_database_id": {
```

as we are recovering the database id specified.

File: azurerm/internal/services/mssql/mssql_database_resource.go [hunk 11]
```diff
@@ -134,6 +134,18 @@ func resourceArmMsSqlDatabase() *schema.Resource {
...
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validate.MsSqlRecoverableDatabaseID,
+			},
+
+			"restorable_dropped_database_id": {
```
Comment (katbyte) [inline azurerm/internal/services/mssql/mssql_database_resource.go:143]: And this 
```suggestion
			"restore_dropped_database_id": {
```

**Review by katbyte — CHANGES_REQUESTED** [review]
@yupwei68 - as i have mentioned before, could we update the property names as suggested? thank you

File: website/docs/r/mssql_database.html.markdown [hunk 12]
```diff
@@ -94,6 +94,10 @@ The following arguments are supported:
 
 * `restore_point_in_time` - (Required) Specifies the point in time (ISO8601 format) of the source database that will be restored to create the new database. This property is only settable for `create_mode`= `PointInTimeRestore`  databases.
 
+* `recoverable_database_id` - (Optional) The ID of the database to be recoveried. This property is only settable and required when `create_mode`=`Recovery`.
+
+* `restorable_dropped_database_id` - (Optional) The ID of the database to be restored. This property is only settable and required when `create_mode`=`Restore`.
```
Comment (katbyte) [inline website/docs/r/mssql_database.html.markdown:99]:
```suggestion
* `restorable_dropped_database_id` - (Optional) The ID of the database to be restored. This property is only applicable when the `create_mode` is `Restore`.
```

File: website/docs/r/mssql_database.html.markdown [hunk 13]
```diff
@@ -94,6 +94,10 @@ The following arguments are supported:
 
 * `restore_point_in_time` - (Required) Specifies the point in time (ISO8601 format) of the source database that will be restored to create the new database. This property is only settable for `create_mode`= `PointInTimeRestore`  databases.
 
+* `recoverable_database_id` - (Optional) The ID of the database to be recoveried. This property is only settable and required when `create_mode`=`Recovery`.
```
Comment (katbyte) [inline website/docs/r/mssql_database.html.markdown:97]:
```suggestion
* `recoverable_database_id` - (Optional) The ID of the database to be recovered. This property is only applicable when the `create_mode` is `Recovery`.
```

File: website/docs/d/mssql_server.html.markdown [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+* `resource_group_name` - (Required) The name of the Resource Group where the Microsoft SQL Server exists.
+
+## Attributes Reference
+
+In addition to the Arguments listed above - the following Attributes are exported: 
+
+* `id` - The ID of the Microsoft SQL Server.
+
+* `administrator_login` - The administrator's login name for the new server.
```
Comment (katbyte) [inline website/docs/d/mssql_server.html.markdown:40]:
```suggestion
* `administrator_login` - The server's administrator login name.
```

File: azurerm/internal/services/mssql/mssql_server_data_source.go [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+						},
+						"tenant_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+
+			"restorable_dropped_database_id": {
```
Comment (katbyte) [inline azurerm/internal/services/mssql/mssql_server_data_source.go:70]: There are multiple IDs here so this should be plural 
```suggestion
			"restorable_dropped_database_ids": {
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Couple more comments from a second pass

**Comment by yupwei68** [comment]
@katbyte @jackofallops Sorry for my previous misunderstandings. I have pushed the changes. Please continue reviewing.

**Review by katbyte — APPROVED** [review]
Thanks @yupwei68 - LGTM 👍

**Pull request merged**